### ISSUE TITLE
Maintenance Supply Requester (MSR) has wrong circuit wiring and logic

### ISSUE THREAD

**Issue by luevano** [Opinionated-Blueprints/10-Books-Full-of-Rails#85]
### Have you checked the following resources?

- [X] [Wiki](https://github.com/Opinionated-Blueprints/10-Books-Full-of-Rails/wiki)
- [X] [Discussions Forum](https://github.com/Opinionated-Blueprints/10-Books-Full-of-Rails/discussions)
- [X] [Open and Closed Issues](https://github.com/Opinionated-Blueprints/10-Books-Full-of-Rails/issues)

### Version of the Book

v2.21.1

### Bug Description

The circuit wiring is completely wrong, and some of the logic as well.

### How to reproduce the problem

Just place the blueprint and you can notice some things:
- The arithmetic combinators for the logistic/constructor robot amounts is not hooked up with the roboport signal.
- The output signals for the train limit and the dispatch were hooked to the wrong wiring.
- Some descriptions don't match the intent of the circuit.
- Some others that I already forgot.

I don't have an example of the original blueprint with the items in the network to show how it is wrong. But I worked on it last night trying to fix it and I think I got it right. Some of the things I did that I remember:
- Rewired most of the circuits, tried to keep it simple.
- Only read the roboport signals and not the network contents (for the arithmetic combinators), because this was causing a doubling of the output signal that the inserters were catching (so if the network needed 300 constructors, the signal would be 600).
- Added/changed some of the descriptions to the best of my knowledge.
- For the quarter request I did the best that I could, not sure how this was intended to work but I think I got it right.
  - I'm using the quarter request as a threshold to know if it makes sense to call a new Maintenance Supply Train (MST). To avoid sending trains even when there is just 1 wall missing, for example.

Also, as a side note, the Wall Maintenance Roboport (WMR) has the wrong default for the robot request group, should be 20 constructors and 10 logistics as per the MSR blueprint, this is minor and a super easy fix on the user end though, but maybe its worth adding a note somewhere and maybe have the default come similar to the MSR blueprint.

The "corrected" MSR blueprint, please tweak as needed:

```
0eNrNWW1v2zYQ/iuEgALbILl6sSQ7QPdlwLAPXTd0G4YhNQxaom2iEqmSVBwvyH/fkbLebCW2nBTdhzY0dffwxOM9d0c9WKusJIWgTFk3D1ZKZCJooShn1o31R1kU2R4pgSlDO5plaEVQgrOMpEhxJMiXkkhFhES7LWFIVuJYIbUl+ql5iKgEtYzvUOi+QXzdqKUI57xkCmGWGo3OOhnBdwRxlpAaVRCcbIlEnuu+mXxin9jPXKCUl6uMOFuCU22S1pfIQeRekQOmVFi/zMSyLZpwJq2b2wdL0g3DmX5fhnMCL9q8iaPXUNYjSLOU3Fs33uPCtghTVFFSKZsf+yUr8xURIGDXIFKvv9kqB/5ksF7BJa028sECJCeYBba1h0EYzWGBlAqSVM+nj/YJrj8C1xuBG1yOG89H4E5H4MYjcMMRuOEI3GgE7hi/xSNwx/htdjluNMZv8wZXwzmH0Bg6Zf6k2d94EvaRPX8A2nMb7JyktMwdkoGCoIlT8IycWySCRYZQ24grsJT0jjiF4Hc0baJ3aKujPq5tARkowbPlimzxHeVCiyZUJCVVS3iWGv0lYRgIJrVu1jiTZNCcNlBXZfbZAQYiApjknBXxWSskUcs1zTS/WjdKlATkS0mO504takNcM57CTDkJz1eUYcWH7YovsUtWzpb9MTBiTZW21ZjWm23OLWfE2UH6ABu+lDgDo2GacZGbEwc2FlgYG2+sd2ai1FnJC104SDm+Xx4mfNd19XsfVmgdsMGKjMbuQcPvDnK7kZuSOaoUgqhx+D24liIzDIfkSsC+vb2daLlyncFftRV89zXWiTovwlInp2zstrvPwbcUWtDiVZFnPWRHcWcjQC59zTVaQi3KvHjRrvdOz3k2fbVX6FCs4CtecKHGwh+hd8HbeM34hkoFL6BXUa+4VUcMKErDVtcsExxtVNDfqGmH/NfrTgZ6rTdpQ3pFNy/x+LOrRGcyxovWgQJaR1kBcn8D+aNfoUqH+hzr4r7qMKzHBaRWq8jwnohlvwf5wFHOBTQHW8xQMNcFffJZQvuhwIyh7Ne6ZI2lej4fx8dVQbdYej47N5WBzsL2aemgZddUSMjjx93G0bGHZSGdgl5HUu0LLXlHhSrNdtc51Eg4/1iPRx740dI7+Gz+vSLatO9OSg7b4ndECCi4lsYZYPa/RHdKA84IR1eAcdivRYZQoxEuDp9xsaair+TjAd65ws9/XuPnKynvxb5u0zYWVG2h/gA3nyk83Y6z+76ZDbumhW7dIIf8cGZn9W3CNR6p9XhBBD5QlANSvFRFecX6XbOXjKgdF5/N+4j23G0EIaxuQY5MPtUxYo2S6RGeJNefcJaUGVTNEt0e7kAWyEG3VJFcIspQldUWk0/sA4HqmupLGfOmEmEgZXIPvSjCd9A56khBR3qg9rtx+ZFaJQaMrppbpfZS6LuSrcsMTqG+ZTpMy+8ng1zf1nPm7seBFqN4viOOLyJ5g7bUaAUZG666batuopYZzamqvdiduxzyvYn3wy0WTN8mPOPinTdzbWiDbOhXFsdJ9fatkVlUbnx3dLe1MJGfVe8J/7mT2A1nMx9icRp783AaTMPYDTwfAnNjHs+mgRfEwSyYRu58PguD0Asj/XilHxvdMPBmUeTGQeTOgjko6KsYrHliyGvzoery9CrAbXj7iYZUEJwuzVla5jwFPNAwc4b0NGGptnM+bENLb080saYUOb2Nc68gt0518VRP/Vp09rHDZXUJV5XhXaL6YYionihKnuSMX/gO5ZjtUa2BQL6KaziI0HCmmgD03atu9dGBnxCYAWpt5Ve7fjCwfe+K7Y762/3Nckm7/860v/9vr04UlzA4pHhdBXWu2IF//5LVXT1lCYSGJNVF+1YQueVZWt3bw5R2HvzLqZSUbSqCHvaMf4Vnwm8dCP6IQBiq3M4HQ1frawREcBFntjfV/wfOnI4u/qNZ56QMF/9+21KkJKlufc/UmDXi9MIW74B7dBC7v27HcwSU0Xu1heCC07SoT9/x56gzIJDAk89VS1Dsqw57uRY8X1IGYHWF+HQv/Z4oaQ5i99xVX92qD260oQ791a4pw4aPZHS1I8Jv6AgI1oMbTjura/3y/gU++a0gTHY/VWontIw8VAxPjKU72D5t5y1keQ8KQdtb2Hrs67HfjOf1ONTzoR57ej6u5v1qPNXjQOPEFU4A85CFjcwU5oFBg2rc4od63qvmzXha6YZaJqplfD1/GId6bGyItLxf6cY2rBZVumCK7R/wZ1omrGRmGieq7JzrsWtkfG0/kKMZGzs1DuyOpjh9adZ83LYtaFEIeNM6vX9CH+saGcQgUKVxDRS58+kcKuI48N0AyOg/+VN1/A==
```

### What did you expect?

The blueprint to work.

### What happened instead?

The blueprint didn't work.

**Comment by Bocian-1** [comment]
Incredible how all of those glaring issues keep slipping past me. I even make testing setups and it still slips by.

> For the quarter request I did the best that I could, not sure how this was intended to work but I think I got it right.
I'm using the quarter request as a threshold to know if it makes sense to call a new Maintenance Supply Train (MST). To avoid sending trains even when there is just 1 wall missing, for example.
You got it right, the station should open if any of the requested items is below 1/4 of the requested amount. There is a hidden addition with the wire from the provider's chest combining signals of its contents with the output of the middle arithmetic combinator that calculates the quarter of the threshold. The division is `/-4` instead of `/4` so it ends up negative but then is brought back up to positive through this hidden addition If the items in the chest are above the threshold. If they're not, the signal remains negative and that's a call for a supply run.
Well, at least that's how it was intended, but clearly I broke it again.

**Comment by Bocian-1** [comment]
>Also, as a side note, the Wall Maintenance Roboport (WMR) has the wrong default for the robot request group, should be 20 constructors and 10 logistics as per the MSR blueprint, this is minor and a super easy fix on the user end though, but maybe its worth adding a note somewhere and maybe have the default come similar to the MSR blueprint.

True for construction robots, but it's better not to request logistic bots to WMR. They will always go to MSR for items to bring back to buffers so ideally they'd stay as close to that as possible to minimise response time.
Now that I think of it more, it would be good to request that all the logistic bots stay at the two MSR roboports

**Comment by luevano** [comment]
> Incredible how all of those glaring issues keep slipping past me. I even make testing setups and it still slips by.

Hey no issues, these things happen. Could also be due to how I "fixed" it initially, as the roboport signal was missing. I only noticed when the train was making loops extremely fast and without moving any item, but it looked fine when I hooked the first signal.

> The division is `/-4` instead of `/4` so it ends up negative but then is brought back up to positive through this hidden addition If the items in the chest are above the threshold. If they're not, the signal remains negative and that's a call for a supply run. Well, at least that's how it was intended, but clearly I broke it again.

Yeah, I experimenting with both positive and negative, and left the one that was easier to hook up in my opinion, but should be easy to swap around.

> True for construction robots, but it's better not to request logistic bots to WMR. They will always go to MSR for items to bring back to buffers so ideally they'd stay as close to that as possible to minimise response time. Now that I think of it more, it would be good to request that all the logistic bots stay at the two MSR roboports

Didn't think about that, just thought it was a missing config on the WMR. But I wouldn't send them all to the two MSR roboports, because what if I have a very long wall with a lot of WMR? then the two MSR roboports would be saturated with logistic robots; unless the logic for the logistic robots is changed to leave at least one stack available for the constructor roboports to propagate.

I haven't seen any issues with the current setup that I described, as there is a buffer anyways in each WMR.

**Comment by Bocian-1** [comment]
Fixed on v2.21.2.
I noticed your implementation changed how the threshold works. Instead of opening the station when any item was below 1/4 of request, it would open it when any item was below request minus threshold, so below 1-1/4 = 3/4

I also added a decider combinator to only set the inserter filter to items actually present on the train so it wouldn't stop working when the filters were set to items already emptied out of the train (such as if bots already took some out) and be unable to take out the rest.

The check for the request being fulfilled to make the train leave early is also gone as for the same reason of bots taking stuff out during unloading it would rarely be satisfied and train would leave on the inactivity check anyways

> Didn't think about that, just thought it was a missing config on the WMR. But I wouldn't send them all to the two MSR roboports, because what if I have a very long wall with a lot of WMR? then the two MSR roboports would be saturated with logistic robots; unless the logic for the logistic robots is changed to leave at least one stack available for the constructor roboports to propagate.

It's only the logistic bots that would be sitting at the MSR, not construction as the only logistics within the wall networks are from the provider chests at MSRs to WMR buffer chests. They won't move items between buffer chests anyway. It's better to keep them near the only providers in the network so they react faster to requests from buffers.
